src/monitor.ts

```diff
@@ -0,0 +1,425 @@
+    // constructor cannot be asynchronous... And since parsing the configuration
+    // can potentially block, we don't want to parse that data synchronously
+    // either. This is a compromise.
+    this.configPath = configPath;
+    for (const node of nodes) {
+      this.nodesStatus.push(this.configToInitialNodeStatus(node));
+    }
+
+    if (typeof heartbeatInterval !== 'undefined') {
+      this.heartbeatInterval = heartbeatInterval;
+    } else {
+      this.heartbeatInterval = 5000;
+    }
+
+    if (typeof failureTolerance !== 'undefined') {
+      this.failureTolerance = failureTolerance;
+    } else {
+      this.failureTolerance = 1;
+    }
+  }
+
+  /**
+   * Start the embedded HTTP server to process requests on the given port.
+   */
+  public start(port: number) {
+    this.timer = this.createPingTimer();
+    this.ping();
+
+    this.server = http.createServer(this.app);
+    this.server.listen(port);
+  }
+
+  /**
+   * Stop the HTTP server.
+   */
+  public stop() {
+    if (typeof this.timer !== 'undefined') {
+      clearInterval(this.timer);
+      this.timer = undefined;
+    }
+
+    if (typeof this.server !== 'undefined') {
+      this.server.close();
+      this.server = undefined;
+    }
+  }
+
+  private validateNetVersionResponse(data: any): INetVersionResponse {
+    if (!data.hasOwnProperty('id')) {
+      throw new Error('response does not contain id field');
+    }
+
+    if (typeof data.id !== 'number') {
+      throw new Error('id is not a number');
+    }
+
+    if (!data.hasOwnProperty('result')) {
+      throw new Error('response does not contain result field');
+    }
+
+    return data as INetVersionResponse;
+  }
+
+  /**
+   * Generates the initial status for a node given a monitored node's
+   * configuration.
+   */
+  private configToInitialNodeStatus(config: IMonitoredNodeConfig): IMonitoredNodeStatus {
+    const result: IMonitoredNodeStatus = {
+      address: config.address,
+      alive: false,
+      lastUpdate: new Date(),
+      lastResponseId: 0,
+      failureCount: 0,
+    };
+
+    if (typeof config.reviveCmd !== 'undefined') {
+      result.reviveCmd = config.reviveCmd;
+    }
+
+    if (typeof config.reviveArgs !== 'undefined') {
+      result.reviveArgs = config.reviveArgs.split(/\s+/);
+    }
+
+    return result;
+  }
+
+  /**
+   * Append the configurations to the configuration jsonline file.
+   */
+  private appendConfigData(nodes: IMonitoredNodeConfig[]) {
+    const file = fs.createWriteStream(this.configPath);
+
+    nodes.forEach(node => {
+      const configLine = JSON.stringify(node);
+      file.write(configLine + '\n');
+    });
+
+    file.end();
+  }
+
+  private setupRouting() {
+    // GET /status returns the current status of all known instances.
+    this.app.get('/status', (req, res) => {
+      res.send(this.nodesStatus);
+    });
+
+    // Parse all request body as JSON, ignoring content-type, for the /add
+    // route.
+    this.app.use('/add', bodyParser.json({
```

Comment:
is there a reason we want to ignore content type? maybe we should just enforce content type on the request side